### PR TITLE
Make use of postgres --sync-walkeepers in tests/cli

### PR DESCRIPTION
**Pull request by knizhnik**

### CODE REVIEW

File: postgres_ffi/src/pg_constants.rs [hunk 1]
```diff
@@ -214,10 +214,9 @@ pub const PGDATA_SUBDIRS: [&'static str; 22] = [
     "pg_logical/mappings",
 ];
 
-pub const PGDATA_SPECIAL_FILES: [&'static str; 4] = [
+pub const PGDATA_SPECIAL_FILES: [&'static str; 3] = [
     "pg_hba.conf",
     "pg_ident.conf",
```
Comment (hlinnaka) [inline postgres_ffi/src/pg_constants.rs:219]: Why this change?

File: pageserver/src/walreceiver.rs [hunk 2]
```diff
@@ -207,7 +207,7 @@ fn walreceiver_main(
 
                 waldecoder.feed_bytes(data);
 
-                while let Some((lsn, recdata)) = waldecoder.poll_decode()? {
+                while let Ok(Some((lsn, recdata))) = waldecoder.poll_decode() {
```
Comment (hlinnaka) [inline pageserver/src/walreceiver.rs:210]: So we now continue processing the WAL stream, if the decoding returns an error once. That seems problematic. Firstly, AFAICS we just ignore the error; at least we should print it to the log. Secondly, what happens next? Will we retry to decode it on the next iteration, and go into infinite loop if the decoding keeps failing? Or will we skip over the WAL record and continue from the next record? Both of those alternatives seem really bad to me.

File: pageserver/src/page_service.rs [hunk 3]
```diff
@@ -357,8 +357,13 @@ impl PageServerHandler {
 
         /* Send a tarball of the latest snapshot on the timeline */
 
-        let req_lsn = lsn.unwrap_or_else(|| timeline.get_last_valid_lsn());
-
+        let req_lsn = match lsn {
```
Comment (hlinnaka) [inline pageserver/src/page_service.rs:360]: Why is this needed? In send_tarball(), the first call to get_page_at_lsn() or list_nonrels() or similar will implicitly call wait_lsn anyway, right? (If they don't do that, I think we should add wait_lsn() calls to those functions instead).

If we want to do an explicit wait_lsn() call, I think that would belong in `send_tarball`, not here.

File: pageserver/src/page_service.rs [hunk 4]
```diff
@@ -468,8 +473,9 @@ impl postgres_backend::Handler for PageServerHandler {
         } else if query_string.starts_with("basebackup ") {
             let (_, params_raw) = query_string.split_at("basebackup ".len());
             let params = params_raw.split(" ").collect::<Vec<_>>();
+			info!("params.len()={},  params[2].len()={}", params.len(), params[2].len());
             ensure!(
-                params.len() == 2,
+                params.len() >= 2,
```
Comment (hlinnaka) [inline pageserver/src/page_service.rs:478]: What's going on here? Does this add a new parameter to `basebackup` command?

**Review by hlinnaka — COMMENTED** [review]
There seem to be a few different changes here:

- Handle CopyFailed message
- Add the explicit wait_lsn() call
- accept an extra basebackup parameter

I'd suggest splitting this up into multiple PRs.

File: walkeeper/src/receive_wal.rs [hunk 5]
```diff
@@ -303,6 +303,7 @@ impl<'pg> ReceiveWalConn<'pg> {
         this_timeline.get().set_info(&my_info);
         /* Need to persist our vote first */
         this_timeline.get().save_control_file(true)?;
+        this_timeline.get().set_info(&my_info);
```
[the remark this answers is not in the document]
Comment (knizhnik) [reply walkeeper/src/receive_wal.rs:306]: Sorry, looks like just merge problem. This call was added by me ~week ago, then remove by Arseny, then restored both by me and Arseny:(

File: postgres_ffi/src/pg_constants.rs (hunk 1 above)
Comment (knizhnik) [reply postgres_ffi/src/pg_constants.rs]: Sorry, reverted.
We need postgresql.conf file to be able to perform recovery: wal_acceptors addresses, timeline and tenant identifiers are specified here. I expected that this file is always generated by control panel/test scripts, so there is no need to include empty postgresq.conf file in tarball. But tests failures shows that this assumption was wrong. So I restoredincluding emoty postgresql.conf in basebackup and just extract titwith -k tar option (do not overwrite existed files.

This is how node is started in @ololobus test:
```
# Remove compute datadir, initialize it using basebackup and start
init_compute () {
    rm -rf zenith_tmp/compute
    mkdir zenith_tmp/compute

    echo "
listen_addresses='0.0.0.0'
password_encryption=md5
max_connections=100
max_wal_senders=10
max_replication_slots=10
hot_standby=on
wal_level=replica
wal_sender_timeout=0
wal_keep_size=256MB
shared_buffers=256MB
wal_log_hints=on
fsync=off
port=54325
shared_preload_libraries='zenith'
synchronous_standby_names='walproposer'
wal_acceptors='127.0.0.1:6501,127.0.0.1:6502,127.0.0.1:6503'
zenith.zenith_tenant='${TENANT_ID}'
zenith.zenith_timeline='${TIMELINE}'
zenith.page_server_connstring='host=127.0.0.1 port=6400'" >> zenith_tmp/compute/postgresql.conf

    sh -c "cd zenith_tmp/compute && VCL=`PGDATA=zenith_tmp/compute postgres --sync-walkeepers` && psql -p6400 -hlocalhost -Xqtc 'basebackup ${TENANT_ID} ${TIMELINE} ${VCL}' | tar -xk"
}
```

File: pageserver/src/walreceiver.rs (hunk 2 above)
Comment (knizhnik) [reply pageserver/src/walreceiver.rs:210]: Sorry, it was one of my first attempts to address problems with node startup in @ololobus  test.
I will revert it.

File: pageserver/src/page_service.rs (hunk 4 above)
Comment (knizhnik) [reply pageserver/src/page_service.rs:477]: It was just a bug: incorrect  handling of optional third (lsn) parameter of basebackup.
Please notice check `params.len() == 3` few lines later...

File: pageserver/src/page_service.rs (hunk 3 above)
Comment (knizhnik) [reply pageserver/src/page_service.rs:360]: Not all repository methods are waiting for lsn now, for example list_rls/list_nonrels.
In principle it is possible to add wait_lsn here. But it seems to be much more efficient to call wait_lsn at the beginning of basebackup, rather than repeat this check in access to each object.
I do no see something bad in pblishing wait_lsn method.


> There seem to be a few different changes here:
> 
>     * Handle CopyFailed message
> 
>     * Add the explicit wait_lsn() call
> 
>     * accept an extra basebackup parameter
> 
> 
> I'd suggest splitting this up into multiple PRs.

I extracted handling of copy error in separate PR:
https://github.com/zenithdb/zenith/pull/447

I do not think that extracting of  wait_lsn should be done separate PR because it is needed only for basebackup

File: postgres_ffi/src/pg_constants.rs (hunk 1 above)
Comment (ololobus) [reply postgres_ffi/src/pg_constants.rs]: Hm, I wasn't entirely correct with this script. In the cloud control plane currently we do basebackup FIRST and then put config. Anyway, we can handle all scenarios if needed: postgresql.conf exists, empty or not, doesn't exist.

Yet it's still an open question to what extent our 'basebackup' should emulate a real basebackup, I think. Probably we should receive only absolutely necessary data from Pageserver, while all config files and empty directories structure should be created locally by CLI, cloud control plane or whatever frontend is used

File: pageserver/src/page_service.rs (hunk 4 above)
Comment (LizardWizzard) [reply pageserver/src/page_service.rs:477]: It was intentional because there was no usages for this third parameter and no tests for that, so I've guarded this case, but didn't remove the following `params.len() == 3`, and marked it with a `TODO`. If I understand it correctly now tests use this option, so I think you can remove the 'TODO'

File: pageserver/src/page_service.rs (hunk 3 above)
Comment (hlinnaka) [reply pageserver/src/page_service.rs:360]: > Not all repository methods are waiting for lsn now, for example list_rls/list_nonrels.
In principle it is possible to add wait_lsn here. But it seems to be much more efficient to call wait_lsn at the beginning of basebackup, rather than repeat this check in access to each object.
I do no see something bad in pblishing wait_lsn method.

Well, it's a weirdly inconsistent API now. Some functions wait implicitly (get_rel_size, get_page_at_lsn), while others expect that the caller has done the wait already (list_rels, list_nonrels). At the very least that requires comments. But it would be more consistent if they all waited. Or none of them waited.

The base backup code isn't performance critical, a few wait_lsn calls won't make any difference. On performance grounds though, you could argue for removing the wait_lsn calls from all the Timeline interface functions. In principle, the code in page_service.rs could call wait_lsn once, when it receives the first GetPageLsn request, and store that LSN in a local variable. In subsequent requests, it could compare the new request LSN with the previous one, and only call wait_lsn if it has changed. But currently, it doesn't make any difference in practice, there is so much other overhead in every request that the wait_lsn calls don't matter.

**Comment by arssher** [comment]
I did review of `--sync-safekeepers` patch on both safekeeper and Postgres side, fixed several things, added comments and rebased the branches. Mine changes are mostly described by this commit msg:
> 
>     Adapt safekeepers to --sync-safekeepers walproposer mode.
>     
>     1) Do epoch switch without record from new epoch, immediately after recovery --
>     --sync-safekeepers mode doesn't generate new records.
>     2) Fix commit_lsn advancement by taking into account wal we have locally --
>        setting it further is incorrect.
>     3) Report it back to walproposer so he knows when sync is done.
>     4) Remove system id check as it is unknown in sync mode.
> 

It think it `--sync-safekeepers` mode and it is usage are ok now to be merged. The problem is that 1) even current `create_node_fixes` branch contains unrelated important commit `Correctly handle LSN parameter in BASEBACKUP command`, which is being reworked in pr #496. 2) Changed `test_restart_compute` still doesn't pass in 100% cases (but it does sometimes, which is an improvement) and it shouldn't: we also need to a) advance `commit_lsn` only to record boundary (implemented by Konstantin in `safe_flush_ptr` branch) (without it we basically corrupt WAL during compute restart) and b) most probably pr #505 -- without WAL truncation `--sync-safekeepers` might never finish as it doesn't generate new records.

We'll see how better to proceed with this. Right now I'll add/review/rebase a) and b) and see how the test is feeling. Maybe #496 gets merged in the meanwhile. We can also try to craft more isolated test specially for `--sync-safekeepers`.

**Comment by arssher** [comment]
#496 merged, I rebased once more. Now this branch contains only `sync-safekeepers` (or directly related) stuff.

File: control_plane/src/compute.rs [hunk 6]
```diff
@@ -342,6 +354,40 @@ impl PostgresNode {
...
+    }
+
+    fn load_basebackup(&self) -> Result<()> {
+        let lsn = if self.uses_wal_proposer {
+            // LSN "0/1000000" means that it is bootstrap and we need to download just
+            // latest data from the pageserver. That is a bit clumsy but whole bootstrap
+            // procedure evolves quite actively right now, so let's think about it again
+            // when things would be more stable (TODO).
+            let lsn = self.sync_walkeepers()?;
+            if lsn == Lsn::from_str("0/1000000").unwrap() {
```
Comment (kelvich) [inline control_plane/src/compute.rs:380]: I think I wrote this, but on a second look I think it would be cleaner to write that as `Lsn(pg_constants::WAL_SEGMENT_SIZE)`

**Review by kelvich — APPROVED** [review]
